### PR TITLE
Fix a memory leak on successful load of CRL

### LINKED ISSUE #23693: A memleak introduced between 3.1.1 and 3.2
I've got a report about a memory leak in curl (originally https://github.com/curl/curl/pull/11531). It happens against 3.2.1 and doesn't happen against 3.1.1, so I consider it as openssl bug. 

```
valgrind ERROR
==309410== 1,795 (248 direct, 1,547 indirect) bytes in 1 blocks are definitely lost in loss record 32 of 32
==309410==    at 0x484280F: malloc (vg_replace_malloc.c:442)
==309410==    by 0x4D77B20: CRYPTO_malloc (mem.c:202)
==309410==    by 0x4D77BD4: CRYPTO_zalloc (mem.c:222)
==309410==    by 0x4C6DFD3: asn1_item_embed_new (tasn_new.c:136)
==309410==    by 0x4C6FB00: UnknownInlinedFun (tasn_new.c:52)
==309410==    by 0x4C6FB00: asn1_item_embed_d2i (tasn_dec.c:366)
==309410==    by 0x4C6FE3F: UnknownInlinedFun (tasn_dec.c:118)
==309410==    by 0x4C6FE3F: ASN1_item_d2i_ex (tasn_dec.c:144)
==309410==    by 0x4D9A4C0: PEM_ASN1_read_bio (pem_oth.c:31)
==309410==    by 0x4DD9C31: X509_load_crl_file (by_file.c:180)
==309410==    by 0x48B6D48: UnknownInlinedFun (openssl.c:3284)
==309410==    by 0x48B6D48: Curl_ssl_setup_x509_store (openssl.c:3437)
==309410==    by 0x48B7445: ossl_bio_cf_in_read (openssl.c:776)
==309410==    by 0x4C73B32: bread_conv (bio_meth.c:121)
==309410==    by 0x4C77C16: bio_read_intern (bio_lib.c:285)
==309410==    by 0x4C77DAA: UnknownInlinedFun (bio_lib.c:311)
==309410==    by 0x4C77DAA: BIO_read (bio_lib.c:303)
==309410==    by 0x4BA1E92: tls_default_read_n (tls_common.c:398)
==309410==    by 0x4BA6B4A: tls_get_more_records (tls_common.c:575)
==309410==    by 0x4BA1099: tls_read_record (tls_common.c:1122)
==309410== 
=======================
```
A reproducer (certs from https://github.com/curl/curl/tree/master/tests/certs):

server: 
`openssl s_server -key stunnel-sv.key -cert stunnel-sv.crt -www`

client:
`valgrind --tool=memcheck --leak-check=yes curl --cacert EdelCurlRoot-ca.crt --crlfile Server-localhost-sv.crl https://localhost:4433/`

The leak doesn't happen without CRL

### PR DESCRIPTION
**Pull request by beldmit**
Fixes #23693

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
I'm not sure the failure is relevant

**Review by shahsb — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Actually - should we have some kind of test for this?

**Comment by t8m** [comment]
Yeah there could be a test that would detect this with MSAN.

**Comment by beldmit** [comment]
Do you have an idea where to add it?

**Comment by bernd-edlinger** [comment]
Simply hijack the test that was introduced by #22885

**Comment by beldmit** [comment]
Tests are added, I lift the hold.

**Comment by beldmit** [comment]
The failures seem irrelevant

File: test/x509_load_cert_file_test.c [hunk 1]
```diff
@@ -36,6 +37,9 @@ static int test_load_cert_file(void)
             goto err;
     }
 
+    if (crl && !TEST_true(X509_load_crl_file(lookup, crl, X509_FILETYPE_PEM)))
```
Comment (mattcaswell) [inline test/x509_load_cert_file_test.c:40]: `crl != NULL`

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/x509_load_cert_file_test.c (hunk 1 above)
Comment (beldmit) [reply test/x509_load_cert_file_test.c]: Done, thanks!

**Comment by beldmit** [comment]
The failure seems irrelevant

**Review by mattcaswell — APPROVED** [review]

**Comment by beldmit** [comment]
Ping @openssl/committers for the 2nd review

**Comment by beldmit** [comment]
Applied to master, doesn't cherry-pick clean to 3.2. Fixup TBD

**Comment by beldmit** [comment]
Thanks for the review!

**Pull request closed without merging**